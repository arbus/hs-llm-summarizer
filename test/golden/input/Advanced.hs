{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FunctionalDependencies #-}

module Advanced where

import Data.Kind (Type)
import qualified Data.Text as T

-- | Phantom type for database tables
data Table a = Table String

-- | Phantom type for table column
data Column a b = Column String

-- | Class for manipulating database tables
class Database db where
  -- | Type of query result
  type QueryResult db :: Type

  -- | Execute a query
  query :: db -> String -> IO (QueryResult db)
  
  -- | Create a new table
  createTable :: Table a -> IO ()
  
  -- | Drop a table
  dropTable :: Table a -> IO ()

-- | Class for orm relationships
class (Database db) => Entity db e | e -> db where
  -- | Get database table for entity
  tableName :: e -> String
  
  -- | List of column names
  columns :: e -> [String]
  
  -- | Primary key
  primaryKey :: e -> String
  
  -- | Get all entities
  getAll :: IO [e]
  
  -- | Get entity by id
  getById :: Int -> IO (Maybe e)
  
  -- | Save entity
  save :: e -> IO e

-- | MySQL database implementation
data MySQL = MySQL { connectionString :: String }

-- | Instance for MySQL database
instance Database MySQL where
  type QueryResult MySQL = [[String]]
  
  query db q = do
    putStrLn $ "Executing on " ++ connectionString db ++ ": " ++ q
    return []
  
  createTable (Table name) = do
    putStrLn $ "Creating table " ++ name
    return ()
  
  dropTable (Table name) = do
    putStrLn $ "Dropping table " ++ name
    return ()

-- | User entity
data User = User
  { userId :: Int
  , userName :: T.Text
  , userEmail :: T.Text
  }

-- | Instance for User entity
instance Entity MySQL User where
  tableName _ = "users"
  
  columns _ = ["id", "name", "email"]
  
  primaryKey _ = "id"
  
  getAll = error "Not implemented"
  
  getById id = do
    putStrLn $ "Getting user with id: " ++ show id
    return Nothing
  
  save user = do
    putStrLn $ "Saving user: " ++ T.unpack (userName user)
    return user