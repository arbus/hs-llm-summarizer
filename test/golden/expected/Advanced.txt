-- Module: test/golden/input/Advanced.hs

 | Phantom type for database tables

data Table a = Table String

 | Phantom type for table column

data Column a b = Column String

 | Class for manipulating database tables

class Database db where
        type QueryResult db :: Type
        
        query :: db -> String -> IO (QueryResult db)
        
        createTable :: Table a -> IO ()
        
        dropTable :: Table a -> IO ()

 | Class for orm relationships

class (Database db) => Entity db e | e -> db where
        tableName :: e -> String
        
        columns :: e -> [String]
        
        primaryKey :: e -> String
        
        getAll :: IO [e]
        
        getById :: Int -> IO (Maybe e)
        
        save :: e -> IO e

 | MySQL database implementation

data MySQL = MySQL{connectionString :: String}



 | User entity

data User = User{userId :: Int, userName :: T.Text,
                 userEmail :: T.Text}

